core/src/main/scala/kafka/server/KafkaApis.scala

```diff
@@ -894,50 +894,84 @@ class KafkaApis(val requestChannel: RequestChannel,
-      val unauthorizedGroupResponse = new OffsetFetchResponse.PartitionData(OffsetFetchResponse.INVALID_OFFSET, "", Errors.GROUP_AUTHORIZATION_FAILED.code)
-      val results = offsetFetchRequest.partitions.asScala.map { topicPartition => (topicPartition, unauthorizedGroupResponse)}.toMap
-      new OffsetFetchResponse(results.asJava)
-    } else {
-      val (authorizedTopicPartitions, unauthorizedTopicPartitions) = offsetFetchRequest.partitions.asScala.partition { topicPartition =>
-        authorize(request.session, Describe, new Resource(auth.Topic, topicPartition.topic))
-      }
-      val unknownTopicPartitionResponse = new OffsetFetchResponse.PartitionData(OffsetFetchResponse.INVALID_OFFSET, "", Errors.UNKNOWN_TOPIC_OR_PARTITION.code)
-      val unauthorizedStatus = unauthorizedTopicPartitions.map(topicPartition => (topicPartition, unknownTopicPartitionResponse)).toMap
-
-      if (header.apiVersion == 0) {
-        // version 0 reads offsets from ZK
-        val responseInfo = authorizedTopicPartitions.map { topicPartition =>
-          val topicDirs = new ZKGroupTopicDirs(offsetFetchRequest.groupId, topicPartition.topic)
-          try {
-            if (!metadataCache.contains(topicPartition.topic))
-              (topicPartition, unknownTopicPartitionResponse)
-            else {
-              val payloadOpt = zkUtils.readDataMaybeNull(s"${topicDirs.consumerOffsetDir}/${topicPartition.partition}")._1
-              payloadOpt match {
-                case Some(payload) =>
-                  (topicPartition, new OffsetFetchResponse.PartitionData(payload.toLong, "", Errors.NONE.code))
-                case None =>
-                  (topicPartition, unknownTopicPartitionResponse)
+      // reject the request if not authorized to the group
+      if (!authorize(request.session, Read, new Resource(Group, offsetFetchRequest.groupId)))
+          new OffsetFetchResponse(Errors.GROUP_AUTHORIZATION_FAILED, offsetFetchRequest.partitions, header.apiVersion)
+      else {
+        val isAllPartitions = offsetFetchRequest.isAllPartitions // could be true only when header.apiVersion >= 2
+
+        val (partitions, authorizedPartitions, unauthorizedPartitions, groupOffsets) =
+          if (!isAllPartitions) {
+            // the request is for specific partitions offsets (all api versions)
+            val partitions = offsetFetchRequest.partitions.asScala.toList
+            val (authorized, unauthorized) =
+              partitions.partition { partition => authorize(request.session, Describe, new Resource(auth.Topic, partition.topic)) }
+            (partitions,
+                authorized,
+                unauthorized,
+                None) // we are not interested in all offsets in the group here
+          } else {
+            // the request is for all group offsets (api version 2 and later)
+            val groupOffsets = coordinator.handleFetchOffsets(offsetFetchRequest.groupId, None)
```

Comment:
I think we need to call `isAllPartitions` upfront anyway because we need to make sure `offsetFetchRequest.partitions` is not null before starting to filter.